library/math/dual.lisp

```diff
@@ -0,0 +1,166 @@
+;;;; dual.lisp
+;;;;
+;;;; dual numbers to enable automatic differentiation
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+    (:use
+     #:coalton
+     #:coalton-library/builtin
+     #:coalton-library/classes
+     #:coalton-library/functions
+     #:coalton-library/math/arith
+     #:coalton-library/math/elementary)
+  (:export
+   #:Dual
+   #:primal-part
+   #:dual-part))
+
+(in-package #:coalton-library/math/dual)
+(named-readtables:in-readtable coalton:coalton)
+
+(coalton-toplevel
+
+  (define-type (Dual :t)
+    "Representation of a Dual number in the form (a + bE)."
+    (Dual :t :t))
+  
+  (declare primal-part (Dual :t -> :t))
+  (define (primal-part (Dual r _)) r)
+
+  (declare dual-part (Dual :t -> :t))
+  (define (dual-part (Dual _ d))
+    d)
+
+  ;; base-part to get the base of a given log
+  (declare base-part (Dual :t -> :t))
+  (define (base-part (Dual b _))
+    b)
+
+  (declare exponent-part (Dual :t -> :t))
+  (define (exponent-part (Dual expt _))
+    expt)
+  
+  (define-instance (Eq :t => Eq (Dual :t))
+    
+    (define (== (Dual a b) (Dual p q))
+      "Check that the given Dual numbers are the same."
```

Comment:
I don't think any of the *instance* doc strings are necessary *unless* the implementation is doing something special, unexpected, or noteworthy. I'd delete these personally.